components/event-system/EventHost.js

```diff
@@ -0,0 +1,47 @@
+import React from 'react'
+import eventQueue from '.'
+
+const EVENT_AWAIT_NEXT = Symbol('EVENT_AWAIT_NEXT')
+const componentRegistry = {}
+
+const EventHost = () => {
+  const [event, setEvent] = React.useState(EVENT_AWAIT_NEXT)
+  if (event === EVENT_AWAIT_NEXT) {
+    if (eventQueue.hasNext()) {
+      setEvent(eventQueue.next())
+    } else {
+      eventQueue.eventQueueWatchOnce(() => {
+        setEvent(eventQueue.next())
+      })
+    }
+    return null
+  }
+  if (event) {
+    if (!event.data.duration) {
+      event.data.duration = 1000
+    }
+    setTimeout(() => setEvent(EVENT_AWAIT_NEXT), event.data.duration)
+    if (componentRegistry[event.type]) {
+      const Component = componentRegistry[event.type]
+      return <Component {...event.data} />
+    }
+    return (
+      <>
+        <code>{'missing component for event type '}{event.type}{' please do EventHost.register(\''}{event.type}{'\',Component)'}</code>
+        <h1>
+          {'did you mean to put one of the following?'}
+        </h1>
+        <ul>
+          {Object.keys(componentRegistry).map(key => (<li key={key}><code>{key}</code></li>))}
+        </ul>
+      </>
+    )
+  }
+  return null
+}
+
+EventHost.register = (type, component) => {
```

Comment:
I don't like that the `componentRegistry` is global. Multiple instances of the `EventHost` would use the same registry, which could be an issue. Instead of the static `register` method, I'd prefer that `EventHost` take a `components` prop. That would dramatically improve the reusability of `EventHost`.

As an example of where this may come into play, we could theoretically have 2 instances of `EventHost` — one for the popups in the middle of the screen, then a separate instance for a smaller persistent list of events along the bottom of the overlay. These two registries may use separate components for each list, in which case we'd need to push 2 separate events (`bit` and `bits-persistent`) to the same queue to accomplish the goal of using two different components.

Reply:
@trezy with this current setup its not really designed to allow for multiple systems to pull from it as the pull operation removes data from the queue

Reply:
Right, that's what I'm saying we should change. We should be able to have multiple queues and multiple `EventHost` components. I think it's *very* likely that there's going to be a use for multiple queues, so the `EventHost` needs to support that, too.